src/library/dao/trackdao.cpp

```diff
@@ -265,32 +264,26 @@ void TrackDAO::saveTrack(TrackPointer track) {
 }
 
 void TrackDAO::saveTrack(TrackInfoObject* pTrack) {
-    DEBUG_ASSERT_AND_HANDLE(pTrack != NULL) {
-        return;
-    }
-    //qDebug() << "TrackDAO::saveTrack" << pTrack->getId() << pTrack->getInfo();
-    TrackId trackId(pTrack->getId());
-    if (trackId.isValid()) {
-        // Track has already been stored in database
-        if (pTrack->isDirty()) {
-            //qDebug() << this << "Dirty tracks before clean save:" << m_dirtyTracks.size();
-            //qDebug() << "TrackDAO::saveTrack. Dirty. Calling update";
-            updateTrack(pTrack);
+    DEBUG_ASSERT(nullptr != pTrack);
```

Comment:
Well, each and every place where a pointer gets dereferenced is a potential 
crash. If we clutter the code with lots of null checks + handling code in 
every function the actual code becomes unreadable. Explicit checks are 
needed at subsystem or component boundaries, but not for every function.

This argument is a pointer, just because we don't allow to use references 
for mutable objects. It will and must never be null, documented by the 
assertion. Passing a mutable reference would be more expressive here ;)

Another option would be to declare the dirty flag in TrackInfoObject as a 
mutable member.

On 01/23/2016 09:50 PM, RJ Ryan wrote:
>
> In src/library/dao/trackdao.cpp 
> <https://github.com/mixxxdj/mixxx/pull/872#discussion_r50625356>:
>
> > @@ -265,32 +264,26 @@ void TrackDAO::saveTrack(TrackPointer track) {
> >  }
> >
> >  void TrackDAO::saveTrack(TrackInfoObject* pTrack) {
> > -    DEBUG_ASSERT_AND_HANDLE(pTrack != NULL) {
> > -        return;
> > -    }
> > -    //qDebug() << "TrackDAO::saveTrack" << pTrack->getId() << pTrack->getInfo();
> > -    TrackId trackId(pTrack->getId());
> > -    if (trackId.isValid()) {
> > -        // Track has already been stored in database
> > -        if (pTrack->isDirty()) {
> > -            //qDebug() << this << "Dirty tracks before clean save:" << m_dirtyTracks.size();
> > -            //qDebug() << "TrackDAO::saveTrack. Dirty. Calling update";
> > -            updateTrack(pTrack);
> > +    DEBUG_ASSERT(nullptr != pTrack);
>
> BTW I would prefer to keep "_AND_HANDLE" for things that would easily 
> prevent a crash.
>
> It may be impossible today but you never know what tomorrow will bring :).
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/872/files#r50625356>.
>